codalab/server/bundle_manager.py

```diff
@@ -581,10 +581,9 @@ def _schedule_run_bundles(self):
         self._bring_offline_stuck_running_bundles(workers)
         self._fail_on_too_many_resources()
         self._acknowledge_recently_finished_bundles(workers)
-
-        # Schedule, preferring user-owned workers.
-        self._schedule_run_bundles_on_workers(workers, user_owned=True)
-        self._schedule_run_bundles_on_workers(workers, user_owned=False)
+        
+        # Schedule bundles to run on workers
+        self._schedule_run_bundles_on_workers(workers)
```

Comment:
My understanding of `user_owned` variable is used to schedule jobs that are requested to run on personal workers first. However, after reading the logic inside `self._schedule_run_bundles_on_workers()`, I noticed that it always loop through the staged bundles fetched from the database without actually scheduling those `user_owned` bundles first. Then there are several unnecessary loops that happened before the real jobs start. I remove `user_owned` logic and created a new list named with `ordered_staged_bundles` to make sure those user_owned bundles will be scheduled first.

Reply:
The `user_owned` here refers to the workers, not the bundles. So, the desired logic is that for any bundle submitted by a user, we should first try to schedule that bundle on a worker that also belongs to that user before trying any of the Codalab workers. This is regardless of whether the user requested a queue or not in the bundle.

I think there's room to refactor this like you started, but we need to keep the logic of preferring user-owned workers over codalab workers for all bundles. Maybe you can try to move that logic to the sort and filter workers function.

But I think it's also fine to keep this part unchanged for the sake of getting this PR finished sooner.

Reply:
Oh, I see. I originally thought `user_owned` means the bundles that were scheduled to run on personal workers, but it's good to get it cleared now. 
I also had the same thought for the desired logic. The logic that I changed here is to save one round of for loop. With the logic of sorting on `staged_bundle.metadata.request_queue, it puts the jobs that were requested to run on personal workers first and then the jobs on public shared workers, which I think is more intuitive in a way that we just need to go through all the `staged` jobs once. But if we don't want to do this right now, I can revert it back. Then we can at least deploy other parts asap.

Reply:
@bkgoksel If you see my above comments, please let me know!

Reply:
Hmm, so `request_queue` should not matter here. We want a given user's bundles to run on their own workers whenever that's possible, regardless of whether `request_queue` is `None` or not on the bundle. In other words, setting `request_queue` doesn't mean scheduling the bundle to only run on your own workers.

What determines whether a run runs on a user-owned worker or public worker is solely whether there's a user-owned worker available at the time we're trying to run that bundle.

As a result, the thing to do here is not sorting the bundles by the `request_queue` parameter. I think the ideal solution would be something like for each staged bundle: 1. get all user owned workers and public workers. 2. sort and filter these workers, and while sorting give first precedence to user-owned workers. 3. In the end, if your top worker is NOT user owned, enforce the parallel run quota.

Reply:
But I think just as a way to move things forward, it might be a good idea to just make a tiny PR where the `user_owned_workers` method does a deepcopy and merge that in, so that prod can be unblocked.
Feel free to make a separate PR for that and tag me on it as well.

Reply:
@bkgoksel I see. I just updated the code. Please check and let me know if it looks good!